### DISCUSSION TITLE
is solid docs createEffect saying promise is not part of microtask?

### DISCUSSION THREAD

**Discussion by dante01yoon** [solidjs/solid#1232] [General]
during reading docs [createEffect section](https://www.solidjs.com/docs/latest/api#createeffect),
it says `If you want to wait for the first execution to occur, use [queueMicrotask](https://developer.mozilla.org/en-US/docs/Web/API/queueMicrotask) (which runs before the browser renders the DOM) or await Promise.resolve() or setTimeout(..., 0) (which run after browser rendering).`

so, according to explanation, queueMicrotask runs before browser renders the DOM,
and `Promise.resolve().then(someCallback)` runs after browser renders the DOM. 

but isn't it both callback function in `queueMicrotask` and promises are kinds of microtask?

so I guess docs should be edited as both tasks runs before browser renders the DOM.

I checked both Promise.resolve.then, queueMicrotask runs before browser renders the DOM using codes below
```js
// assume this code is in a component function, so is part of a rendering phase
const [count, setCount] = createSignal(0);

// this effect prints count at the beginning and when it changes
createEffect(() => console.log("count =", count()));
// effect won't run yet
console.log("hello");
setCount(1); // effect still won't run yet
setCount(2); // effect still won't run yet
Promise.resolve().then(() => alert("promise"))
queueMicrotask(() => {
  // now `count = 2` will print
  alert("microtask");
  setCount(3); // immediately prints `count = 3`
  console.log("goodbye");
});

// --- overall output: ---
// hello
// count = 2
// alert promise
// alert microtask
// count = 3
// goodbye
```

**Comment by eyelidlessness** [comment]
`Promise.resolve().then(cb)` is synchronous, and should be equivalent to `queueMicrotask(cb)`. If I recall correctly, that was even how the proposal for `queueMicrotask` described it. In both cases, `cb` is added to the end of the current task’s queue in the event loop.

It’s the right hand side of an `await` that will occur after the DOM renders (even if the right hand side isn’t a `Promise`!), because it asynchronously yields to the event loop. The docs could maybe be more clear on this, but even more general docs like MDN don’t exactly do a great job clearly explaining the various event loop scheduling semantics in a consolidated way.

**Comment by lxsmnsyc** [comment]
> and Promise.resolve().then(someCallback) runs after browser renders the DOM.

I think this is a misinterpretation, where the remark `(which run after browser rendering)` should only refer to `setTimeout`, but yes it could've been phrased differently.
